### PR TITLE
Initial creation of workflows building block

### PR DESCRIPTION
**Pull request by RyanLettieri**
Signed-off-by: Ryan Lettieri <ryanLettieri@microsoft.com>

# Description

Initial Creation of the workflows building block using temporal

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Related Proposal: https://github.com/dapr/dapr/issues/4576

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by berndverst** [comment]
Adding the `do-not-merge` label to make sure no maintainer merges this before the 1.9 release.

File: workflows/responses.go
Comment (cgillum) [inline workflows/responses.go:5]: Why `string` and not `time.Time`?

**Review by cgillum — COMMENTED** [review]
Still making my way through this PR. Here's my first batch of comments.

File: tests/certification/workflows/temporal/temporal_test.go
Comment (RyanLettieri) [reply tests/certification/workflows/temporal/temporal_test.go:30]: Certification test was added by mistake and was not ready. It is going to be removed in this PR and added in at a later date

File: workflows/responses.go
Comment (RyanLettieri) [reply workflows/responses.go:5]: I was thinking that returning it as a string could provide more flexibility for formatting purposes or even flexibility for other backends if they return time as a string. If that doesn't make sense though, I'll gladly make the change to use time.Time

File: workflows/README.md
Comment (cgillum) [inline workflows/README.md:3]: Should we add a link to the proposal so that readers can get more context on this feature area?

File: workflows/README.md
Comment (cgillum) [inline workflows/README.md:7]:
```suggestion
A compliant workflow needs to implement the `Workflow` interface included in the [`workflow.go`](workflow.go) file.
```

File: workflows/requests.go
Comment (cgillum) [inline workflows/requests.go:16]: Same comment I left on the other PR, but I suggest using a name such as `WorkflowReference` instead of `WorkflowInfo`. I think this will help with clarity.
```suggestion
	WorkflowReference WorkflowReference `json:"workflow_reference"`
```

File: workflows/responses.go
Comment (cgillum) [reply workflows/responses.go:5]: Will there be any Go code that consumes this type? My concern is about whether 1) any Go code needs to consume this field and 2) whether it may be displayed somewhere. In the latter case, it would be a little awkward to have different components display the creation time using different formats. Not a big deal, but I wonder if it could cause problems, especially if we need to actually consume these values somewhere, like in the Dapr Dashboard. Having a strict type/format might also make it easier to display start times in a sorted manner.

File: workflows/temporal/temporal.go
Comment (cgillum) [inline workflows/temporal/temporal.go:100]: Should we add a log statement for this case, like we do for the other error cases?

**Review by cgillum — COMMENTED** [review]
Added a few more comments.

I noticed that .github/infrastructure/conformance/temporal/server/dynamicconfig/docker.yaml is empty. I'm not sure how conformance tests are supposed to work, but is this expected? Seems like there's another one for the certification test as well (also, I thought you were removing the certification tests?).

Also, there's a binary file in this PR: .github/infrastructure/conformance/temporal/worker/worker. Is this supposed to be there?

File: .github/infrastructure/conformance/temporal/worker/go.mod
Comment (RyanLettieri) [reply .github/infrastructure/conformance/temporal/worker/go.mod:25]: There seems to be a compatibility issue when running with that version:
*workflowServiceProxyServer does not implement workflowservice.WorkflowServiceServer (missing GetWorkerBuildIdOrdering method)

Comment (cgillum) [reply .github/infrastructure/conformance/temporal/worker/go.mod:25]: Weird. But I guess it's not a blocker for this PR. We can revisit this later.

File: workflows/responses.go
Comment (RyanLettieri) [reply workflows/responses.go:5]: Updated the temporal code to output the time using a time standard (RFC3339). It is still a string

**Review by cgillum — DISMISSED** [review]
:shipit:

**Comment by berndverst** [comment]
Conformance: https://github.com/dapr/components-contrib/actions/runs/3055926528
Certification: https://github.com/dapr/components-contrib/actions/runs/3055926532

**Review by berndverst — APPROVED** [review]

**Comment by RyanLettieri** [comment]
This PR is being held open until the end of 1.9 in-case there're some hidden broken changes. Furthermore, since the API changes are not pushed into master, this isn't a strong need to have this merged into 1.9 The PR will be updated with the latest changes from master to ensure that it is not out of date. Once the code-freeze for 1.9 is complete, this can be merged into 1.10

**Review by berndverst — CHANGES_REQUESTED** [review]
Please resolve conflicts and then we can get this merged into master.

File: tests/certification/go.mod
Comment (cgillum) [inline tests/certification/go.mod:8]: I'm just now noticing this, but why are we using such an old version of the Go SDK in this PR? I would have expected us to use the latest v1.5.x version.

**Comment by berndverst** [comment]
I'll merge once @cgillum can do another quick check and give an approval.

File: workflows/requests.go
Comment (cgillum) [inline workflows/requests.go:23]: TaskQueue is a Temporal-specific concept and thus shouldn't belong outside of the temporal package. But is the `CreateWorkerRequest` being used anywhere?

File: workflows/requests.go
Comment (cgillum) [inline workflows/requests.go:28]: Same question, is this being used anywhere? I couldn't find any references.

File: workflows/temporal/temporal.go
Comment (cgillum) [inline workflows/temporal/temporal.go:102]: Now that I've had a bit more time to digest Go, I'm wondering if it would be more conventional to return a custom error instead of logging like this. The expectation is that the upstream caller takes care of logging for us:
```go
return fmt.Errorf("error terminating workflow: %w", err)
```
Same comment for all other places where we log and then return an error.

**Review by cgillum — COMMENTED** [review]
I noticed there's a binary in this PR under /.github/infrastructure/conformance/temporal/worker named `worker`. Is that supposed to be checked into GitHub?

File: workflows/requests.go
Comment (berndverst) [reply workflows/requests.go:23]: Follow up question: Is there a design doc for what the workflow interface actually should look like?

At this level there should generally only be things matching the protos and relevant helpers that apply to all components of the workflow type.

File: workflows/temporal/temporal.go
Comment (berndverst) [reply workflows/temporal/temporal.go:102]: I agree. Since `Terminate` is part of the Dapr runtime, the Runtime actually should log any error to the logger.

In this place we should just `return fmt.Errorf`

File: workflows/requests.go
Comment (berndverst) [reply workflows/requests.go:23]: @RyanLettieri what you put here should match the proto and changes in https://github.com/dapr/dapr/pull/5158

**Comment by berndverst** [comment]
> I noticed there's a binary in this PR under /.github/infrastructure/conformance/temporal/worker named `worker`. Is that supposed to be checked into GitHub?

The `worker` binary definitely needs to be removed @RyanLettieri. Seems your Docker container builds the binary from the source file anyway.

File: tests/certification/workflows/temporal/temporal_test.go
Comment (cgillum) [inline tests/certification/workflows/temporal/temporal_test.go:31]: Does this file need to be removed? My previous comment was marked as resolved, and I was expecting this would have been removed. I'm not sure what the current expectation is. We don't appear to be actually testing anything here.

File: workflows/temporal/temporal.go
Comment (cgillum) [inline workflows/temporal/temporal.go:31]: It looks like this const isn't being used. Given the value, it looks like it doesn't belong in production code either.

**Review by cgillum — APPROVED** [review]
One more minor thing, but otherwise this looks good to me.

**Review by cgillum — APPROVED** [review]

File: go.mod
Comment (berndverst) [inline go.mod:125]: can you please clean this up and make sure there are only 2 `require` blocks here? The first block is the direct depedencies, the second one is the indirect ones.

File: go.mod
Comment (berndverst) [inline go.mod:340]: Don't mix direct and indirect dependencies in different blocks please.

**Review by berndverst — CHANGES_REQUESTED** [review]
Your workflow conformance test is not running in CI/CD.

Please make sure you first run your conformance test locally.

Should be something like
`go test -v -tags=conftests -count=1 ./tests/conformance -run="TestWorkflowConformance/temporal"`

Only you will know the specific based on your PR.

image

**Comment by berndverst** [comment]
I was able to verify the workflow locally with:

```
docker-compose -f ./.github/infrastructure/docker-compose-temporal.yml -p temporal up -d
go test -v -tags=conftests -count=1 ./tests/conformance -run="TestWorkflowsConformance/temporal"
=== RUN   TestWorkflowsConformance
=== RUN   TestWorkflowsConformance/temporal
=== RUN   TestWorkflowsConformance/temporal/init
time="2022-10-27T12:34:28.51554-07:00" level=debug msg="Temporal init start" instance=Bernds-MacBook-Pro.local scope=testLogger type=log ver=unknown
2022/10/27 12:34:28 INFO  No logger configured for temporal client. Created default one.
=== RUN   TestWorkflowsConformance/temporal/start
time="2022-10-27T12:34:28.526645-07:00" level=info msg="Start test running..." instance=Bernds-MacBook-Pro.local scope=workflowsTest type=log ver=unknown
time="2022-10-27T12:34:28.526677-07:00" level=debug msg="starting workflow" instance=Bernds-MacBook-Pro.local scope=testLogger type=log ver=unknown
time="2022-10-27T12:34:28.540806-07:00" level=debug msg="getting workflow data" instance=Bernds-MacBook-Pro.local scope=testLogger type=log ver=unknown
time="2022-10-27T12:34:33.546585-07:00" level=debug msg="getting workflow data" instance=Bernds-MacBook-Pro.local scope=testLogger type=log ver=unknown
time="2022-10-27T12:34:33.55706-07:00" level=debug msg="terminating workflow" instance=Bernds-MacBook-Pro.local scope=testLogger type=log ver=unknown
time="2022-10-27T12:34:33.576683-07:00" level=debug msg="getting workflow data" instance=Bernds-MacBook-Pro.local scope=testLogger type=log ver=unknown
time="2022-10-27T12:34:33.585117-07:00" level=info msg="Start test done." instance=Bernds-MacBook-Pro.local scope=workflowsTest type=log ver=unknown
--- PASS: TestWorkflowsConformance (5.07s)
    --- PASS: TestWorkflowsConformance/temporal (5.07s)
        --- PASS: TestWorkflowsConformance/temporal/init (0.01s)
        --- PASS: TestWorkflowsConformance/temporal/start (5.06s)
PASS
ok  	github.com/dapr/components-contrib/tests/conformance	5.112s
```

**Comment by berndverst** [comment]
@RyanLettieri please open a Docs issue to create the new Workfows section in the Docs, and then list the temporal component as Alpha (due to the fact the building block is Alpha)

**Review by berndverst — APPROVED** [review]

**Comment by RyanLettieri** [comment]
> @RyanLettieri please open a Docs issue to create the new Workfows section in the Docs, and then list the temporal component as Alpha (due to the fact the building block is Alpha)

@berndverst would this previously created issue suffice: https://github.com/dapr/docs/issues/2798 ?

**Pull request merged**
